### PR TITLE
feat: SSB-2 add endpoints for password recovery and validate password…

### PR DESCRIPTION
**Pull request by AntonioLok**
# Description

This PR adds 2 new endpoints which will be used for to reset a user password if they have forgot it.

1-  1st endpoint: POST endpoint `users/forgot-password`this endpoint sends a recovery email with a link to reset a user password. This link will contain a hash/token that is created and stored in our DB for a specific user. If the user requesting to reset his/her password does not exist, we will return a 404. Otherwise if it exists, we will return a 200 and an email will be sent to the user.

2-  2nd endpoint: GET endpoint `users/reset/:token`this endpoint serves to validate the token that is passed in the request parameter. We check against the token (if there is any) and the expiration time that we have stored for the user requesting a password recovery.

# Test
0 - checkout this branch `git checkout SSB-2-create-endpoints-for-password-recovery-and-validate-link-hash`
1 - Install dependencies `npm i`
2 - Set the required environment variables `JWT_SECRET`, `EMAIL_ADDRESS`, `EMAIL_PASSWORD` by creating a `.env` file in the main directory and set your secret in that file ie. `JWT_SECRET=HELLO`. For the email address and email password, please use: 
```
EMAIL_ADDRESS=os.recovery.pw@gmail.com
EMAIL_PASSWORD=random.password
```

The email address and password is used by the nodemailer module to send emails to other users with the recovery password link.
3 - Run the app using `npm run start`.
4 - Now, using the command line, send the following HTTP requests to test our 2 new endpoints:

To test our endpoints, we need an account first (SKIP this if you have an account):
- Endpoint for registering an acc `http://localhost:8000/users/register`:
    -  method: POST
    -  body parameters:
        - `email` type: string
        - `password` type: string

sample command:
`curl -d "email=antonio@gmail.coml&password=password" -X POST http://localhost:8000/users/register`.

The response should be:

```
 {"statusCode":201,"statusText":"CREATED","data":null}
```

Now that you have your email, we can do our test!

-  1st endpoint `/users/forgot-password`
    -  method: POST 
    -  body parameters:
        - `username` type: string

sample command:
`curl -d "username=antoniochan030@gmail.com" -X POST http://localhost:8000/users/forgot-password`

Plz do not use the above sample command. Replace the username with the email you used to register before. Note that an email will be sent to that email address so if you were to try the above sample command, I will be getting the email on my personal email.

The response should be:

```
 {"statusCode":200,"statusText":"SUCCESS","data":null}
```
Now try with an email that does not exist ie. username=antoniochan04324324234cz30@gmail.com
`curl -d "username=antoniochan04324324234cz30@gmail.com" -X POST http://localhost:8000/users/forgot-password`

The response should be :

```
{"statusCode":404,"statusText":"NOT_FOUND","message":"The requested resource or endpoint was not found."}
```

-  2nd endpoint `/users/reset/:token`
    -  method: GET

Check your inbox. An email should have been sent to you with the password reset link. It should look something like this: `http://localhost:8080/reset-password/21352db78fb5eac7827087ad54e5e0fd13419291` where `21352db78fb5eac7827087ad54e5e0fd13419291` is the token.

Now send a GET request to `/users/reset/:token` with your token 

sample command:
`curl -X GET http://localhost:8000/users/reset/21352db78fb5eac7827087ad54e5e0fd13419291`

Sample response should be something similar to this:
```
{"statusCode":200,"statusText":"SUCCESS","data":{"username":"antoniochan030@gmail.com"}}
```

Now try the same command after 1 hour (which is the expiration time for the token) or try with a invalid token ie. invalidtoken123
`curl -X GET http://localhost:8000/users/reset/invalidtoken123`

Response should be;
```
{"statusCode":400,"statusText":"BAD_REQUEST","message":"The request was malformed or missing a required parameter."}
```
# Next to do
Ticket to be created - endpoint to update user password

### CODE REVIEW

File: src/views/routes/users.js [hunk 1]
```diff
@@ -61,16 +65,61 @@ router.post('/login', async (req, res) => {
   }
 });
 
+router.get('/:username', async (req, res) => {
```
Comment (AntonioLok) [inline src/views/routes/users.js]: TODO - CREATE TICKET: This was an old endpoint which is used by the front-end as an async validation to check if the user has already registered with the username passed as the request parameter. The logic in the frontend should be updated to accommodate these changes.

**Comment by howardhenry** [comment]
Hey, I've dropped a comment on the JIRA ticket regarding the scope of this feature. Take a look and let me know before I start reviewing the PR

**Comment by AntonioLok** [comment]
> Hey, I've dropped a comment on the JIRA ticket regarding the scope of this feature. Take a look and let me know before I start reviewing the PR

Checking now. Sorry about that!

**Comment by AntonioLok** [comment]
Closing PR as I am updating the scope of the ticket and I need to make some changes to this PR. @howardhenry

**Pull request closed without merging**